src/oracle/monitor.js

```diff
@@ -0,0 +1,429 @@
+            let lastRound = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+
+            //get feeds last observed round from state
+            let lastObservedRound = state["values"].hasOwnProperty(feed) ? state["values"][feed]["round"] : 0
+
+            //if round is bigger than last observed and the offer didn't fail
+            if (lastRound > lastObservedRound && !currentOffer["status"].hasOwnProperty("error")) {
+
+                //if id is bigger than last offer id in state, set it
+                lastResults["last_offer_id"] = id
+
+                let price = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["unitPrice"]) / amountsIn[feed]
+
+                //fill results variable
+                lastResults["values"][feed] = {
+                    price: price,
+                    id: id,
+                    round: lastRound
+                }
+                state = lastResults
+
+                //get latest feed price
+                let feedPrice = await queryPrice(feed)
+                //update metrics
+                updateMetrics(oracleDetails["oracleName"], oracle, feed, price, id, feedPrice, lastRound)
+
+            }
+        }
+    }
+
+    //loop through balances
+    for (var i = 0; i < offersBalances.balances.length; i++) {
+        let currentBalance = offersBalances.balances[i]
+
+        var brand = currentBalance.brand.iface.split(" ")[1]
+        if (brand.includes("BLD") || brand.includes("IST")) {
+            var value = Number(currentBalance.value)
+            updateBalanceMetrics(oracleDetails["oracleName"], oracle, brand, value)
+        }
+    }
+
+    return lastResults["last_offer_id"] != lastOfferId ? lastResults : state
+}
+
+/**
+ * Function to read the latest monitoring state from file
+ * @returns latest monitoring state
+ */
+const readMonitoringState = () => {
+    //try to read from file
+    try {
+        return readJSONFile(STATE_FILE)
+    } catch (err) {
+        //if it fails, initialise and save
+        let initialState = {}
+
+        for (let oracle in oracles) {
+            initialState[oracle] = {
+                "last_index": 0,
+                "values": {}
+            }
+        }
+
+        //save to file
+        saveJSONDataToFile(initialState, STATE_FILE)
+        return initialState
+    }
+}
+
+/**
+ * Main function to monitor
+ */
+export const monitor = async () => {
+
+    //create interval
+    setInterval(async () => {
+
+        //read monitoring state
+        let state = readMonitoringState()
+
+        //for each oracle
+        for (let oracle in oracles) {
+
+            //check if there is state for oracle
+            if (!(oracle in state)) {
+                state[oracle] = {
+                    "last_offer_id": 0,
+                    "values": {}
+                }
+            }
+            console.log("ORACLE STATE", oracle, state[oracle])
```

Comment:
Why is this log important?

Reply:
Removed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/327efaef5faa89f92c783d6d0bb45c1d809d6757